### DISCUSSION TITLE
Add Maturin to an existing rust library as a subcrate/workspace

### DISCUSSION THREAD

**Discussion by tgross35** [PyO3/maturin#995] [Q&A] [answered]
Hello all,

Is there a good way to add a python package to an existing rust library, without changing that library (avoiding pyo3 as a dependency)? I feel like this should be possible, maybe with [cargo workspaces](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html), but have yet to figure out how to do this.

I have an existing crates.io project that has a rust library, and a CLI binary. I'd like to create a python module for PyPi that uses this library, and keep it contained in the same package so it uses the same version numbering. This will require some wrapping both rust-side and python-side, which I don't really want to ship with the main crate, both to avoid confusion and to help keep dependencies slimmer.

My thought was to use a directory structure like the following:

```bash
some_project/
├── Cargo.toml
├── bindings_py/          # Main directory for rust project
│   ├── Cargo.toml        # Don't really want this file to avoid managing "version"
│   │                     # Currently has dependency `some_project = { path = "../" }`
│   ├── docs/             # Sphinx docs for the python package
│   ├── pysrc/            # This is what should be called in python by `import some_project`
│   │   │                 # Includes all python-side wrapping
│   │   ├── __init__.py   # This should import the builds and do necessary python-side wrapping
│   │   └── __init__.pyi  # Typing goes here, I think (?)
│   ├── src/
│   │   └── lib.rs        # Rust-side wrapping for the library
│   └── tests/            # Pytest integration tests for the python & rust combined
│       └── some_python_test.py  
├── pyproject.toml
├── README_pypi.md        # Also unsure how to specify the README to use for pypi
├── README.md
├── src/
│   ├── bin/
│   │   └── bin.rs        # This creates the CLI executable that ships with the package
│   ├── lib.rs            # Main package library
│   └── something.rs      # Support for main package library
├── target/
│   ├── debug
│   └── wheels
└── tests/                # Standard rust integration tests
    └── rust_integration_test.rs
```

But I have been struggling with getting Maturin to use the version from the top level `Cargo.toml` and use the correct subcrate for `maturin build`/`maturin develop`. Also getting tools like `cargo check` and `cargo fmt` to recognize the submodule, but that's not necessarily related to Maturin.

Is something like this possible, or is there a better structure for creating mixed rust/python libraries?

**Comment by messense** [comment, marked as answer]
> Cargo.toml        # Don't really want this file to avoid managing "version"

It's not possible today since Cargo requires the `version` field, see https://github.com/rust-lang/cargo/issues/9829. But I think you can use something like [cargo-workspaces](https://github.com/pksunkara/cargo-workspaces) or [cargo-release](https://github.com/crate-ci/cargo-release) to manage your crate versions.

You can't not avoid `Cargo.toml` since you need to specify Rust dependencies.

> Also unsure how to specify the README to use for pypi

https://maturin.rs/metadata.html

**Reply by tgross35** [reply]
Thanks for the advice with cargo-release, it has gotten me something workable.

My working repo at this point with GH actions, for anyone interested: https://github.com/pluots/stringmetrics/tree/41958f697eb32847454e8bafadcbe679ba30fdde
